pkg/verify/signature.go

```diff
@@ -146,56 +147,40 @@ func verifyEnvelopeWithArtifact(verifier signature.Verifier, envelope EnvelopeCo
 	if err = limitSubjects(statement); err != nil {
 		return err
 	}
-
-	var artifactDigestAlgorithm string
-	var artifactDigest []byte
-
-	// Determine artifact digest algorithm by looking at the first subject's
-	// digests. This assumes that if a statement contains multiple subjects,
-	// they all use the same digest algorithm(s).
+	// Sanity check (no subjects)
 	if len(statement.Subject) == 0 {
 		return errors.New("no subjects found in statement")
 	}
-	if len(statement.Subject[0].Digest) == 0 {
-		return errors.New("no digests found in statement")
-	}
 
-	// Select the strongest digest algorithm available.
-	for _, alg := range []string{"sha512", "sha384", "sha256"} {
-		if _, ok := statement.Subject[0].Digest[alg]; ok {
-			artifactDigestAlgorithm = alg
-			continue
-		}
-	}
-	if artifactDigestAlgorithm == "" {
-		return errors.New("could not verify artifact: unsupported digest algorithm")
+	// determine which hash functions to use
+	hashFuncs, err := getHashFunctions(statement)
```

Comment:
What do you think about maintaining a multi hasher per subject? With the current approach, we have to hash every subject with all algorithms, which is potentially inefficient. For example, if we have subject A with digests sha512 and sha384, and subject B with digest sha256, then we only need to compute 3 total digests, though with this approach, we'd compute 6.

Reply:
This code still only supports calculating the digests for a single artifact. In your example, it would compute two digests (sha512 and sha256) for the same artifact and compare against both subjects in the attestation.

If we add support for #363, we would need to calculate four total digests (both sha512 and sha256) for each input file, and compare against the subject digests.

Reply:
You're right, I was thinking we specify the subject name we're verifying, rather than searching for a given digest. The optimization I specified would require knowing the subject name, which isn't something checked currently.